source/ui/memory-component-presenter.cpp

```diff
@@ -42,12 +42,33 @@ MemoryComponentPresenter::~MemoryComponentPresenter() {
 
 void MemoryComponentPresenter::onMemoryChanged(std::size_t address,
                                                std::size_t length) {
-  emit dataChanged(this->index(address / 1),
-                   this->index(address + length - 1));//  8bit
-  emit dataChanged(this->index(address / 2),
-                   this->index(address + length - 1));// 16bit
-  emit dataChanged(this->index(address / 4),
-                   this->index(address + length - 1));// 32bit
+  // calculate region for (max) 64bit memory cells
+  int start = address - (address % (64 / 8));
+  int end = start + (length - (length % (64 / 8)) + (64 / 8)) - 1;
+
+  // region should not exceed real memory size
+  if (end >= _memorySize) end = _memorySize - 1;
+  if (start <= 0) start = 0;
+
+  // if the memory that is hold in cache is changed, invalidate cache
+  // new value intersects with the beginning of the cached region
+  bool overlap_beginning =
+      (address <= _memoryCacheBaseAddress + _memoryCacheSize &&
+       address + length >= _memoryCacheBaseAddress + _memoryCacheSize);
+  // new value intersects with the ending of the cached region
+  bool overlap_ending = (address <= _memoryCacheBaseAddress &&
+                         address + length >= _memoryCacheBaseAddress);
+  // new value is completly inside cached region
+  bool overlap_middle =
+      (address >= _memoryCacheBaseAddress &&
+       address + length <= _memoryCacheBaseAddress + _memoryCacheSize);
+
+  if (overlap_beginning || overlap_middle || overlap_ending) {
+    _memoryCacheValid = false;
+  }
```

Comment:
couldn't this be written as `start < _memoryCacheBaseAddress + _memoryCacheSize && end > _memoryCacheBaseAddress` ?